### ISSUE TITLE
[MEAR-267] assembly.xml contains incorrect references to modules

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-ear-plugin#431]
**[Leonid Rozenblyum](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leokom)** opened **[MEAR-267](https://issues.apache.org/jira/browse/MEAR-267?redirect=false)** and commented

SCENARIO:

Create a EAR project with maven-ear-plugin 3.0.0

execute mvn ear:ear

EXPECTED:

assembly.xml contains reference to the jar/war equivalent to their physical names inside

the EAR

(e.g. if the jar is named tryEar-ejb-1.0-SNAPSHOT.jar then assembly.xml reference would be:

> `<module>`
> \<ejb>tryEar-ejb-1.0-SNAPSHOT.jar\</ejb>
>  \</module>

(this worked in 2.10.1)

ACTUALLY:
assembly.xml contains reference

> \<module>
> > >  \<ejb>com.leokom-tryEar-ejb-1.0-SNAPSHOT.jar\</ejb>
> > >  \</module>

 

Due to this difference - JBoss/WildFly cannot deploy the EAR.

(it's easy to reproduce: you may create a ear project from some standard ones from maven-archetypes and change maven-ear-plugin version to 3.0.0).

 

UPDATE: Sorry, maybe it's a bug in M2E-WTP plugin of Eclipse. I tried this scenario in standalone mode without Eclipse - and assembly.xml is consistent with the jar files

---

**Affects:** 3.0.0

**Issue Links:**
- [MEAR-153](https://issues.apache.org/jira/browse/MEAR-153) Skinny Modules -- not just WARs
 (_**"is fixed by"**_)

**Remote Links:**
- [GitHub Pull Request #19
](https://github.com/apache/maven-ear-plugin/pull/19)

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

The changes are mentioned on the page of the plugin as [change for 3.0.0](https://maven.apache.org/plugins/maven-ear-plugin/). Apart from that can you check if it's really a bug in M2E ? If you checked you can give a feedback in the comments...Thanks.

**Comment by jira-importer** [comment]
**[Leonid Rozenblyum](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leokom)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) I checked it and indeed:
* the bug exists in Eclipse (so most likely belongs to M2E or M2E-WTP, not sure)
* the bug doesn't exist in standalone maven execution.

Should I move the ticket to M2E?

**Comment by jira-importer** [comment]
**[Leonid Rozenblyum](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leokom)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) In any case the behaviour introduced by 3.0.0 is a problem now.

How is it possible force the dependent jar manifests respect the new naming of jars?

 

so <project.ear> contains <project.jar> which should have a correct Manifest referring to the libraries in the 'ear'.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Have you checked the docs to change the naming ?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Do you have an example project for that so you can attach to this jira issue so I can take a look....

**Comment by jira-importer** [comment]
**[Marat Abrarov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abrarovm)** commented

Hi [Leonid Rozenblyum](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leokom),

Won't workaround in [pull request #1](https://github.com/lrozenblyum/MEAR-267-issue-demo/pull/1) of lrozenblyum/MEAR-267-issue-demo project work for you? I understand that it's not solution, e.g. the same EJB JAR may be included into different EARs each having different [outputFileNameMapping](https://maven.apache.org/plugins/maven-ear-plugin/examples/customize-file-name-mapping.html) and the workaround doesn't work in this case.

It looks like with 3.0 version of Maven EAR Plugin default outputFileNameMapping was changed to include group ID. I explicitly defined outputFileNameMapping to look like:

```
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-ear-plugin</artifactId>
    <version>3.0.2</version>
    <configuration>
        ...
        <outputFileNameMapping>@{artifactId}@-@{baseVersion}@@{dashClassifier?}@.@{extension}@</outputFileNameMapping>
    </configuration>
</plugin>
```

when migrated one of my projects from Maven EAR Plugin 2.10.1 to 3.0.2. May be the same solution is more applicable to your needs.

Thank you for detecting and describing this bug.

**Comment by jira-importer** [comment]
**[Marat Abrarov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abrarovm)** commented

I added integration test deploying WildFly 20.0.1.Final (the latest version of WildFly ex JBoss AS) and EAR into that WildFly instance here (is based on [my master branch](https://github.com/lrozenblyum/MEAR-267-issue-demo/pull/1) with [workaround](https://issues.apache.org/jira/browse/MEAR-267?focusedCommentId=17202439&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17202439)): https://github.com/mabrarov/MEAR-267-issue-demo/compare/master...mabrarov:feature/wildfly-managed-it

Test can be executed with simple (tested on JDK 11 and 1.8):

```bash
$ mvn clean verify
```

I added integration test deploying JBoss AS 7.1.1.Final (the oldest version of JBoss AS I was able to run on Oracle JDK 1.7) and EAR into that JBoss AS instance here (is based on my master branch with workaround): https://github.com/mabrarov/MEAR-267-issue-demo/compare/master...mabrarov:feature/jboss-managed-it

Test can be executed with this command (tested on OpenJDK 1.8 + Oracle JDK 1.7 as `jboss.java.home` maven property and on Oracle JDK 1.7):

```bash
$ mvn clean verify -Djboss.java.home=${JAVA7_HOME}
```

where `${JAVA7_HOME`} is location of Oracle JDK 1.7. `jboss.java.home` maven property may be omitted if Maven runs on Oracle JDK 1.7.

Without [workaround](https://issues.apache.org/jira/browse/MEAR-267?focusedCommentId=17202439&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17202439) I get just **warning** and **no errors** on WildFly 20.0.1.Final:

```
00:38:03,488 WARN  [org.jboss.as.server.deployment] (MSC service thread 1-6) WFLYSRV0059: Class Path entry guava-19.0.jar in XXX/MEAR-267-issue-demo/MEAR-267-ear/content/MEAR-267.ear/com.leokom-MEAR-267-ejb-1.0-SNAPSHOT.jar  does not point to a valid jar for a Class-Path reference.
```

and on JBoss AS 7.1.1.Final:

```
00:50:33,947 WARN  [org.jboss.as.server.deployment] (MSC service thread 1-5) Class Path entry guava-19.0.jar in "XXX/MEAR-267-issue-demo/MEAR-267-ear/content/MEAR-267.ear/com.leokom-MEAR-267-ejb-1.0-SNAPSHOT.jar"  does not point to a valid jar for a Class-Path reference.
```

I believe we can decrease priority of this bug, because:
1. this bug doesn't cause error / deployment failure on semi-modern and modern Java EE application servers
2. [workaround](https://issues.apache.org/jira/browse/MEAR-267?focusedCommentId=17202439&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17202439) exists

One more thing I found with JBoss AS/ WildFly is that this Maven EAR Plugin configuration

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-ear-plugin</artifactId>
    <version>3.0.2</version>
    <configuration>
        ...
        <outputFileNameMapping>@{artifactId}@-@{baseVersion}@@{dashClassifier?}@.@{extension}@</outputFileNameMapping>
    </configuration>
</plugin>
```

is not sufficient to solve these warnings - we need to provide "lib/" prefix for class-path entries to make JBoss AS / WildFly happy.

**Comment by jira-importer** [comment]
**[Marat Abrarov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abrarovm)** commented

I implemented possible fix for this bug in [pull request #19](https://github.com/apache/maven-ear-plugin/pull/19). I appreciate if [Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) and [Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo) could help me with review and creation of tests.
